text/0000-postfix-match.md

```diff
@@ -0,0 +1,321 @@
+    .send()
+    .await?
+    .json::<Option<String>>()
+    .await?
+    .as_ref()
+```
+
+```rust
+// prefix match
+match context.client
+    .post("https://example.com/crabs")
+    .body("favourite crab?")
+    .send()
+    .await?
+    .json::<Option<String>>()
+    .await?
+    .as_ref()
+{
+    Some("") | None => "Ferris"
+    x @ Some(_) => &x[1..]
+};
+
+// postfix match
+context.client
+    .post("https://example.com/crabs")
+    .body("favourite crab?")
+    .send()
+    .await?
+    .json::<Option<String>>()
+    .await?
+    .as_ref()
+    .match {
+        Some("") | None => "Ferris"
+        x @ Some(_) => &x[1..]
+    };
+```
+
+While it's a matter of taste, the postfix form is much easier to parse
+in terms of the flow of data. Having to jump back out to the beginning of
+the expression to see the `match` keyword to know what the new expression context is
+can be difficult to read.
+
+While I am not at liberty to share the codebase, the former is equivalent to something
+I see on some of our codebases at work.
+
+## tap (pipelines)
+[guide-level-explanation]: #tap-pipelines
+
+While not the intended use of this proposal, [tap](https://crates.io/crates/tap) is a popular crate to allow more advanced pipeline operations.
+
+```rust
+let val = original_value
+  .pipe(first)
+  .pipe(|v| second(v, another_arg))
+  .pipe(third)
+  .pipe(|v| last(v, another_arg));
+```
+
+This can be written similarly using our postfix match:
+
+```rust
+let val = original_value
+  .match { v => first(v) }
+  .match { v => second(v, another_arg) }
+  .match { v => third(v) }
+  .match { v => last(v, another_arg) };
+```
+
+This avoids the need for a new dedicated pipeline operator or syntax.
+
+### into
+
+Occasionally I've bumped into the difficulty of using `Into` in a postfix setting.
+The tap crate also provides a `Conv` trait to convert in a method chain using turbofish.
+
+```rust
+let upper = "hello, world"
+    .conv::<String>()
+    .tap_mut(|s| s.make_ascii_uppercase());
+```
+
+This can also be emulated with postfix-match
+
+```rust
+let upper = "hello, world"
+    .match { s => String::from(s) }
+    .match { mut s => { s.make_ascii_uppercase(); s } }
+```
+
+## async support
+
+One thing of note is that in option chains you cannot use futures unless you use adapters
+like [`OptionFuture`](https://docs.rs/futures/0.3.21/futures/future/struct.OptionFuture.html).
+
+Using match, you can avoid that by supporting `.await` directly.
+
+```rust
+context.client
+    .post("https://example.com/crabs")
+    .body("favourite crab?")
+    .send()
+    .await
+    .match {
+        Err(_) => Ok("Ferris"),
+        Ok(resp) => resp.json::<String>().await,
+        //         this works in a postfix-match ^^^^^^
+    }
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`X.Y.match {}.Z`
+
+Will be interpreted as
+
+```rust
+match X.Y {}.Z`
+```
+
+I believe this would be the same precedence as `await`.
```

Comment:
This needs to be updated to be a firm statement, not an "I believe".  (And, ideally, a grammar update, though I'm not sure we have a sufficiently-good grammar from which to start in order to do that.)

For `.await` it was quite obvious what the precedence needed to be, since we wanted it used with `?`.  And that's *why* it got the `.`, because `*foo() await?.bar()` was deemed too weird, even though it would have been technically possible for the grammar to be like that.

But with postfix `match` it's not so obvious to me that this is the correct precedence.  `*foo.bar().baz()` is already unfortunate, but I think that
```rust
*foo
    .match {
        Ok(3) => …,
        Ok(100) => …,
        _ => …,
    }
```
makes it *way* worse.

C#'s postfix `switch`, for example, doesn't use the dot, and has [lower precedence](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/#operator-precedence) than the primary expressions, and thus `-x switch {` is `(-x) switch {`.  Whereas in Rust if `.match` has the same precedence as `.await`, then `-x.match { … }` would be `-(x.match { … })`.

It's not obvious to me that that's the precedence that a postfix match *ought* to have, though if it's `.match` then that's probably the one it has to be.  So I'd like to see a Rationale section speaking specifically about precedence and syntax like this.